libwild/src/parsing.rs

```diff
@@ -133,7 +133,8 @@ impl<'data> ParsedInputObject<'data> {
     /// it defines. This is true for archive entries and shared objects for which --as-needed is
     /// true.
     pub(crate) fn is_optional(&self) -> bool {
-        self.input.entry.is_some() || (self.is_dynamic() && self.modifiers.as_needed)
+        !self.modifiers.whole_archive
+            && (self.input.entry.is_some() || (self.is_dynamic() && self.modifiers.as_needed))
```

Comment:
It looks like the parentheses might be in the wrong spot. The --whole-archive flag shouldn't affect shared objects, but it looks like setting --whole-archive here will cause shared objects to become non-optional even if they're --as-needed.